Add Taskfile to build package; Refactor docs tasks.

kirkrodrigues opened the pull request:
# Description
<!-- Describe what this request will change/fix and provide any details necessary for reviewers -->

This PR:
* Adds a [Taskfile](https://taskfile.dev/usage/) to make it easier to build clp_ffi_py
* Refactors the docs Taskfile so that it can be called from the main Taskfile.
* Applies our latest guidelines for Taskfiles to the docs tasks.

NOTE: This PR requires task v3.35.1 for the latest bug fixes.

# Validation performed
<!-- What tests and validation you performed on the change -->

* Validated building each artifact from a clean state:

  ```bash
  task clean && task
  task clean && task wheel
  task clean && task docs:site
  ```

* Validated repeated builds of each artifact-generating task did not result in the artifact being regenerated:

  ```bash
  task clean && task wheel && task wheel
  task clean && task docs:site && task docs:site
  ```

* Validated that after running `task docs:site`, changing any of the following files triggered a rebuild of just the docs:
  * `docs/conf/conf.py`
  * `docs/src/index.md`
  * `docs/tasks.yml`
* Validated that after running `task wheel`, changing any of the following triggered a rebuild of the wheel:
  * `clp_ffi_py/__init__.py`
  * `src/clp` (the version of the clp submodule)
  * `src/clp_ffi_py/error_messages.hpp`
  * `Taskfile.yml`
* Validated that after running `task docs:site`, changing any of the following files from the previous step triggered a rebuild of the wheel and the docs.
* Validated `task clean && task docs:serve` built and served the site.

kirkrodrigues committed:
Add Taskfile to build package; Refactor docs tasks.

kirkrodrigues committed:
Undo temporary change.

kirkrodrigues committed:
Minor fixes.

kirkrodrigues committed:
Add precondition for checkout submodules.

kirkrodrigues commented:
Workflow failures unrelated to the present PR. They'll be addressed in another PR.

junhaoliao reviewed (approved):
It seems we do not checksum all `build/dist` artifacts. e.g.,
```shell
rm -f build/dist/*.tar.gz
task # does not trigger `wheel` rebuild
```

Just want to make sure whether that is intended as I understand the `build/dist/*.whl` file is the final packaged file. Approved assuming that is intended.

junhaoliao commented:
Title: ?
```
Add a main `Taskfile` to make it easier to build clp_ffi_py; Refactor the docs `Taskfile` to make it callable by the main Taskfile. (#46)
```

junhaoliao commented on Taskfile.yml:
```diff
@@ -0,0 +1,107 @@
+      - "rm -rf '{{.OUTPUT_DIR}}'"
+      - |-
+        . "{{.WHEEL_VENV_DIR}}/bin/activate"
+        python3 -m build --outdir "{{.OUTPUT_DIR}}"
```
```suggestion
        python3 -m build --outdir "{{.OUTPUT_DIR}}"
      # Checksum the generated files (this command must be last)
      - |-
        cd "{{.OUTPUT_DIR}}"
        tar cf - {{.CHECKSUM_TAR_BASE_ARGS}} . | md5sum > "{{.CHECKSUM_FILE}}"
```

junhaoliao commented on Taskfile.yml:
```diff
@@ -0,0 +1,107 @@
+      - "{{.TASKFILE}}"
+      - "clp_ffi_py/**/*"
+      - "src/**/*"
+    generates:
+      - "{{.OUTPUT_DIR}}/*.whl"
```
```suggestion
    status:
      - "test -f '{{.CHECKSUM_FILE}}'"
      - >-
        diff
        <(cd '{{.OUTPUT_DIR}}' && tar cf - {{.CHECKSUM_TAR_BASE_ARGS}} . | md5sum)
        "{{.CHECKSUM_FILE}}"
```

kirkrodrigues committed:
Add .tar.gz to wheel's generated files.

junhaoliao reviewed (changes requested):
In fact, hashing of the `*.whl` file does not seem correctly handled. I modified the magic number `PK` of `build/dist/clp_ffi_py-0.0.10-cp310-cp310-linux_x86_64.whl` and running the `task` command did not rebuild the task.

kirkrodrigues committed:
Checksum wheel output manually.

kirkrodrigues commented:
> It seems we do not checksum all build/dist artifacts. e.g.,

I did intentionally leave out the `.tar.gz`, but that is probably not a good idea since the `.tar.gz` can be used to build the package from source. Added the `.tar.gz`.

> In fact, hashing of the *.whl file does not seem correctly handled.

Huh, strange. I would've expected `task` to checksum the file properly by itself. Anyway, switched to the explicit md5 checksum.

junhaoliao reviewed (approved):
Looks good.
